Use find8 and find16 in AdaptiveStringSearcher

Constellation opened the pull request:
#### 31005e5e9299e6a7957bb30a85458a77a719ddad
<pre>
Use find8 and find16 in AdaptiveStringSearcher
<a href="https://bugs.webkit.org/show_bug.cgi?id=274010">https://bugs.webkit.org/show_bug.cgi?id=274010</a>
<a href="https://rdar.apple.com/127894478">rdar://127894478</a>

Reviewed by Mark Lam.

We should just use find8 and find16 in AdaptiveStringSearcher.
Since we have find16, we do not need to use memchr hack here.

* Source/WTF/wtf/text/AdaptiveStringSearcher.h:
(WTF::AdaptiveStringSearcherBase::findFirstCharacter):
(WTF::AdaptiveStringSearcherBase::alignDown): Deleted.
(WTF::AdaptiveStringSearcherBase::getHighestValueByte): Deleted.

Canonical link: <a href="https://commits.webkit.org/278644@main">https://commits.webkit.org/278644@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b8d9e6a6ce6b355d02223df51af90c67c5dae17d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/51161 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/30463 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/3496 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/54418 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/1851 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/36759 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/1525 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/41668 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/53260 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/28095 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/44095 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/22785 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/25417 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/1351 "Passed tests") | [⏳ 🛠 wpe-skia ](https://ews-build.webkit.org/#/builders/WPE-Skia-Build-EWS "Waiting in queue, processing has not started yet") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/44500 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/47397 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/1424 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/56014 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/50663 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/26271 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/1314 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/49066 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/27519 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/44162 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/48205 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/28400 "Built successfully") | | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/59367 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7445 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/27250 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/59367 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
<!--EWS-Status-Bubble-End-->

Achierius commented on Source/WTF/wtf/text/AdaptiveStringSearcher.h:
```diff
@@ -65,51 +65,30 @@ class AdaptiveStringSearcherBase {
+        PatternChar patternFirstChar = pattern[0];
 
-        if (sizeof(SubjectChar) == 2 && !patternFirstChar) {
-            // Special-case looking for the 0 char in other than one-byte strings.
-            // memchr mostly fails in this case due to every other byte being 0 in text
-            // that is mostly ascii characters.
-            for (int i = index; i < maxN; ++i) {
-                if (!subjectPtr[i])
-                    return i;
-            }
```
Does this no longer apply? Or is it just not worth the upfront runtime?

Constellation replied:
This no longer applies since we use find16 instead of memchr.

Achierius commented on Source/WTF/wtf/text/AdaptiveStringSearcher.h:
```diff
@@ -65,51 +65,30 @@ class AdaptiveStringSearcherBase {
-            ASSERT(maxN - pos >= 0);
-            const SubjectChar* charPos = reinterpret_cast<const SubjectChar*>(memchr(subjectPtr + pos, searchByte, (maxN - pos) * sizeof(SubjectChar)));
-            if (charPos == nullptr)
+        if constexpr (sizeof(PatternChar) == 2 && sizeof(SubjectChar) == 1) {
+            if (!isLatin1(patternFirstChar))
```
Why could we handle this case before and not now?

Constellation replied:
Because the previous one was using memchr and a hack for both 16 / 8 bit cases. Now the logic is completely different.

Achierius reviewed (approved):
r=me

ddegazio reviewed (approved):
LGTM!

MenloDorian reviewed (approved):
r=me

Constellation committed:
Use find8 and find16 in AdaptiveStringSearcher
https://bugs.webkit.org/show_bug.cgi?id=274010
rdar://127894478

Reviewed by Mark Lam.

We should just use find8 and find16 in AdaptiveStringSearcher.
Since we have find16, we do not need to use memchr hack here.

* Source/WTF/wtf/text/AdaptiveStringSearcher.h:
(WTF::AdaptiveStringSearcherBase::findFirstCharacter):
(WTF::AdaptiveStringSearcherBase::alignDown): Deleted.
(WTF::AdaptiveStringSearcherBase::getHighestValueByte): Deleted.

Canonical link: https://commits.webkit.org/278644@main
